File: csharp/roman-numerals/RomanNumerals.cs
using System;
using System.Collections.Generic;

public static class RomanNumeralExtension
{
    private static readonly Dictionary<string, int> RomanNumbers = new Dictionary<string, int>
    {
        {"I", 1},
        {"IV", 4},
        {"V", 5},
        {"IX", 9},
        {"X", 10},
        {"XL", 40},
        {"L", 50},
        {"XC", 90},
        {"C", 100},
        {"CD", 400},
        {"D", 500},
        {"CM", 900},
        {"M", 1000}
    };

    public static string ToRoman(this int value)
    {
        if (value == 0)
        {
            return "";
        }
        string roman = "";

        string key = FindNearestNumber(value);

        if (value == RomanNumbers[key])
        {
            roman += key;
        }
        else
        {
            roman += key;
            roman += (value - RomanNumbers[key]).ToRoman();
        }

        return roman;
    }

    private static string FindNearestNumber(int number)
    {
        string nearestKey = "I";

        foreach (var key in RomanNumbers.Keys)
        {
            if (Math.Abs(RomanNumbers[key]) <= number)
            {
                nearestKey = key;
            }
        }

        return nearestKey;
    }
}

File: cpp/roman-numerals/roman_numerals.cpp
#include "roman_numerals.h"
#include <vector>
#include <utility>

namespace roman_numerals
{
    const std::vector<std::pair<std::string, int>> ROMAN_NUMBERS{
        {"M", 1000},
        {"CM", 900},
        {"D", 500},
        {"CD", 400},
        {"C", 100},
        {"XC", 90},
        {"L", 50},
        {"XL", 40},
        {"X", 10},
        {"IX", 9},
        {"V", 5},
        {"IV", 4},
        {"I", 1}};

    std::string find_nearest_number(int number)
    {
        for (const auto &[key, value] : ROMAN_NUMBERS)
        {
            if (value <= number)
            {
                return key;
            }
        }

        return "I";
    }

    std::string convert(int value)
    {
        if (value == 0)
        {
            return "";
        }

        std::string key = find_nearest_number(value);

        std::string roman = key;

        if (value != ROMAN_NUMBERS[0].second)
        {
            int number = 0;

            for (const auto &[k, v] : ROMAN_NUMBERS)
            {
                if (k == key)
                {
                    number = v;
                    break;
                }
            }

            roman += convert(value - number);
        }

        return roman;
    }

} // namespace roman_numerals
